rs/nns/governance/src/storage/neurons.rs

```diff
@@ -343,10 +344,76 @@ where
     /// Returns the next neuron_id equal to or higher than the provided neuron_id
     pub fn range_neurons<R>(&self, range: R) -> impl Iterator<Item = Neuron> + '_
     where
-        R: RangeBounds<NeuronId>,
+        R: RangeBounds<NeuronId> + Clone,
     {
-        self.main.range(range).map(|(neuron_id, neuron)| {
-            self.reconstitute_neuron(neuron_id, neuron, NeuronSections::all())
+        let (first, last) = extract_range_ends(range.clone());
+
+        let hotkeys_range = (first, u64::MIN)..(last, u64::MAX);
+        let ballots_range = (first, u64::MIN)..(last, u64::MAX);
+
+        let followees_range = FolloweesKey {
+            follower_id: first,
+            ..FolloweesKey::MIN
+        }..FolloweesKey {
+            follower_id: last,
+            ..FolloweesKey::MAX
+        };
+
+        let main_range = self.main.range(range.clone());
+        let mut hot_keys_iter = self.hot_keys_map.range(hotkeys_range).peekable();
+        let mut recent_ballots_iter = self.recent_ballots_map.range(ballots_range).peekable();
+        let mut followees_iter = self.followees_map.range(followees_range).peekable();
+        let mut known_neuron_data_iter = self.known_neuron_data_map.range(range.clone()).peekable();
+        let mut transfer_iter = self.transfer_map.range(range).peekable();
+
+        main_range.map(move |(main_neuron_id, abridged_neuron)| {
+            // We'll collect data from all relevant maps for this neuron_id
+
+            let hot_keys = collect_for_neuron_id(
+                &mut hot_keys_iter,
+                main_neuron_id,
+                |((neuron_id, _index), principal)| (*neuron_id, PrincipalId::from(*principal)),
+            );
+
+            let ballots = collect_for_neuron_id(
+                &mut recent_ballots_iter,
+                main_neuron_id,
+                |((neuron_id, _index), ballot_info)| (*neuron_id, *ballot_info),
+            );
+
+            let followees = collect_for_neuron_id(
+                &mut followees_iter,
+                main_neuron_id,
+                |(followees_key, followee_id)| {
+                    (followees_key.follower_id, (*followees_key, *followee_id))
+                },
+            );
+
+            let current_known_neuron_data = collect_for_neuron_id(
+                &mut known_neuron_data_iter,
+                main_neuron_id,
+                |(neuron_id, known_neuron_data)| (*neuron_id, known_neuron_data.clone()),
+            )
+            .first()
+            .cloned();
```

Comment:
To avoid cloned, use pop.

Also, we might want to make sure there are no other elements?